### ISSUE TITLE
Fused compiled-training path silently fails parameter propagation for BatchNorm-heavy CNN models

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#348]
## Summary

When the fused compiled-training path (`CompiledTapeTrainingStep.TryStepWithFusedOptimizer`) engages on a BatchNorm-heavy CNN topology (specifically the GraFPrint paper-faithful 53-layer pyramid: Conv1×1 + BatchNorm + LeakyReLU FFN blocks at channels [64, 128, 256, 512]), `plan.Step()` reports success and `_fusedStepCount` increments correctly, but the parameter updates **do not propagate to the layer-owned trainable tensors**.

Concrete repro: after 20 fused Train calls on a freshly-initialized GraFPrint with a fixed (input, target) pair, the first 10 parameters returned by `network.GetParameters()` are *bitwise identical* to their pre-train values (max |Δ| = 0.0 to fp64 precision). The m/v buffers inside the compiled plan are presumably updating, but the parameter-write hook misses the layer-owned `Tensor<T>` references that `GetParameters()` reads from.

Effects: training tests that depend on parameter motion all fail silently when fused engages:
- `Training_ShouldChangeParameters` — params static
- `Training_ShouldReduceLoss` — loss flat
- `LossStrictlyDecreasesOnMemorizationTask` — loss flat
- `GradientFlow_ShouldBeNonZeroAndFinite` — gradients exist but never reach params
- `TrainingError_ShouldNotExceedTestError` — train MSE ≥ test MSE because trained model = untrained

## How I diagnosed

testconsole harness `grafprint-perf-diag` in https://github.com/ooples/AiDotNet PR #1316 (branch `fix/pr-1290-cluster-4-5`):

\`\`\`
GraFPrint Train: 92.5 ms/iter ⇒ 23.1s for 250 iters  (fused engaged)
After 20 fused Train: first-10-params max |Δ|=0.000000E+000
\`\`\`

vs eager path (which produces correct training but is 13× slower):
\`\`\`
GraFPrint Train: 1230 ms/iter ⇒ 307s for 250 iters  (fused gated off)
After 20 eager Train: first-10-params max |Δ|=2.3e-3
\`\`\`

The gate that keeps fused off is `TryMapToFusedOptimizerConfig` rejecting `AdamOptimizer.UseAdaptiveLearningRate = true` (the framework-wide default).
Forcing the default to `false` engages fused but exposes this propagation bug.

## Likely root cause

The `_cachedParameters` static list in `CompiledTapeTrainingStep<T>` is populated from `CollectDeduplicatedParameters(layers)` on first call and reused thereafter. If the compiled plan's parameter write-back targets a different tensor reference than what the layers actually hold — for example, if BatchNormalizationLayer's `_gamma` / `_beta` allocations are *replaced* (not in-place updated) when the layer transitions from lazy to resolved, the cached list holds dead references and writes go to orphaned tensors.

The NoisyDenseLayer and DuelingCombinationLayer fixes in the parent repo (PR #1316) hit a similar issue on the eager-path ParameterBuffer — they were rebinding fields rather than copying in-place. BatchNormalizationLayer almost certainly has the same pattern.

## Reproducer + suggested fix

1. Land the parameter-buffer-view rebind pattern in `BatchNormalizationLayer.SetTrainableParameters` (non-readonly fields, replace reference instead of `CopyTensorInPlace`).
2. Verify `_cachedParameters` invalidation on model construction / structural change.
3. Add an integration test that asserts `|Δ params| > 0` after a fused Train step on a BN-CNN.

## Why this matters

The fused path's 13× speedup is currently inaccessible to every model that uses BatchNorm + default Adam. Unlocking it on consumer hardware brings paper-faithful deep models (RAPIDFlow, GraFPrint, ConformerFP, anything from the ICASSP/ICRA fingerprint and optical-flow families) inside the xUnit 120-second per-test budget without test-watering — currently those models can only fit by truncating the iteration count (the VGG-style override pattern).

**Comment by ooples** [comment]
## Update: not a Tensors bug — root cause isolated to AiDotNet `ConvolutionalLayer`

The Tensors team was right to push back. I built a minimal repro that bypasses the GraFPrint topology entirely and drives `TryStepWithFusedOptimizer` against single layers directly. The Tensors-side fused optimizer behaves correctly.

### Per-layer verdict on `AiDotNet.Tensors@0.75.5`

| Layer | Fused steps | Loss trajectory | `max |Δ| param` after 20 steps |
|---|---|---|---|
| `DenseLayer` (8→4, Identity activation) | 20 / 20 | 4.65 → 1.56 | **2.0×10⁻¹** ✓ |
| `BatchNormalizationLayer` (rank-2, training mode) | 20 / 20 | 23.9 → 18.9 | **2.0×10⁻¹** ✓ |
| `ConvolutionalLayer` (1×1, 1→4, stride 1) | 20 / 20 | 22.81 → 22.81 (FLAT) | **0.0** ✗ |
| Composite: Conv1×1 → BN → LeakyReLU | 20 / 20 | 1.06M → 0.70M | Conv: 0.0  ✗ , BN: 2.0×10⁻¹ ✓ |

In the composite case the loss drops, but only the BN's γ/β get updated. Conv kernels and biases stay at their initialization values bit-for-bit (max |Δ| = 0.0 to fp32 precision). `ReferenceEquals` between collect-time and post-training `GetTrainableParameters()` returns true for every tensor — there is **no tensor-identity rebind** to blame. The references the caller hands to `GetOrCompileTraining` are the same references after training.

So Tensors is faithfully:
- updating caller-owned `Tensor<T>` storage in place when the optimizer-step writes to a parameter slot, and
- not silently swapping references underneath the caller.

The bug is on the AiDotNet side: `ConvolutionalLayer`'s compiled forward path doesn't connect `_kernels` and `_biases` into the loss-bearing graph node that `cache.GetOrCompileTraining(traceFunc, parameters)` traces. Either:

1. the convolution op the layer's Forward calls is dispatched off-graph (a raw kernel call that doesn't register the weight tensor as a graph leaf), or
2. the weight tensor is referenced by an op the autodiff compiler treats as constant w.r.t. its inputs.

Either way: that's an `AiDotNet.NeuralNetworks.Layers.ConvolutionalLayer` followup, not a Tensors fix.

### Reproducer

Available in `ooples/AiDotNet` on branch `fix/pr-1290-cluster-4-5`:
- `testconsole/FusedPropagationMinRepro.cs` — drives `TryStepWithFusedOptimizer` via reflection on four cases (Dense / Conv / BN / mini-stem). Invocation: `dotnet run --project testconsole/AiDotNetTestConsole.csproj -- fused-min-repro`.

### Disposition

Closing this issue against Tensors — the diagnostic was wrong about where the bug lives. I'll open a new issue against AiDotNet's `ConvolutionalLayer` with the per-layer table above. Apologies for the noise.

**Comment by ooples** [comment]
Closing — diagnostic isolated the bug to AiDotNet ConvolutionalLayer, not Tensors. See followup comment above.
